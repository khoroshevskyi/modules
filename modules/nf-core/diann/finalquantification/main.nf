process DIANN_FINALQUANTIFICATION {
    tag "$meta.id"
    label 'process_high'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://containers.biocontainers.pro/s3/SingImgsRepo/diann/v1.8.1_cv1/diann_v1.8.1_cv1.img' :
        'docker.io/biocontainers/diann:v1.8.1_cv1' }"

    input:
    tuple val(meta), val(ms_files), path(fasta), path(empirical_library), path("quant/")

    output:
    tuple val(meta), path("diann_report.tsv"), emit: main_report, optional: true
    tuple val(meta), path("diann_report.parquet"), emit: report_parquet, optional: true
    tuple val(meta), path("diann_report.manifest.txt"), emit: report_manifest, optional: true
    tuple val(meta), path("diann_report.protein_description.tsv"), emit: protein_description, optional: true
    tuple val(meta), path("diann_report.stats.tsv"), emit: report_stats
    tuple val(meta), path("diann_report.pr_matrix.tsv"), emit: pr_matrix
    tuple val(meta), path("diann_report.pg_matrix.tsv"), emit: pg_matrix
    tuple val(meta), path("diann_report.gg_matrix.tsv"), emit: gg_matrix
    tuple val(meta), path("diann_report.unique_genes_matrix.tsv"), emit: unique_gene_matrix
    tuple val(meta), path("diannsummary.log"), emit: log
    tuple val(meta), path("empirical_library.tsv"), emit: final_speclib, optional: true
    tuple val(meta), path("empirical_library.tsv.skyline.speclib"), emit: skyline_speclib, optional: true
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DIANN_FINALQUANTIFICATION module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''

    """
    # Notes: if .quant files are passed, mzml/.d files are not accessed, so the name needs to be passed but files
    # do not need to pe present.

    diann   --lib ${empirical_library} \\
            --fasta ${fasta} \\
            --f ${(ms_files as List).join(' --f ')} \\
            --threads ${task.cpus} \\
            --temp ./quant/ \\
            ${args}

    cp diann_report.log.txt diannsummary.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        DIA-NN: \$(diann 2>&1 | grep "DIA-NN" | grep -oP "\\d+\\.\\d+(\\.\\w+)*(\\.[\\d]+)?")
    END_VERSIONS
    """

    stub:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DIANN_FINALQUANTIFICATION module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''

    """
    touch diann_report.tsv
    touch diann_report.parquet
    touch diann_report.manifest.txt
    touch diann_report.protein_description.tsv
    touch diann_report.stats.tsv
    touch diann_report.pr_matrix.tsv
    touch diann_report.pg_matrix.tsv
    touch diann_report.gg_matrix.tsv
    touch diann_report.unique_genes_matrix.tsv
    touch diannsummary.log
    touch empirical_library.tsv
    touch empirical_library.tsv.skyline.speclib

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        DIA-NN: \$(diann 2>&1 | grep "DIA-NN" | grep -oP "\\d+\\.\\d+(\\.\\w+)*(\\.[\\d]+)?")
    END_VERSIONS
    """
}
