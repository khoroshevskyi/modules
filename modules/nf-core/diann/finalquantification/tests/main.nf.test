nextflow_process {

    name "Test Process DIANN_FINALQUANTIFICATION"
    script "../main.nf"
    process "DIANN_FINALQUANTIFICATION"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "diann"
    tag "diann/finalquantification"

    test("proteomics - final quantification") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [experiment_id:'test_exp'],
                    ['RD139_Narrow_UPS1_0_1fmol_inj1.mzML'],
                    file(params.modules_testdata_base_path + 'proteomics/diann/empirical_library.speclib', checkIfExists: true),
                    [file(params.modules_testdata_base_path + 'proteomics/diann/individual_analysis.quant', checkIfExists: true)],
                    file(params.modules_testdata_base_path + 'proteomics/diann/REF_EColi_K12_UPS1_combined_subset_100.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path("${process.out.log[0][1]}").exists() },
                { assert snapshot(
                    process.out.report_stats,
                    process.out.pr_matrix,
                    process.out.pg_matrix,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("proteomics - final quantification - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of([
                    [experiment_id:'test_exp'],
                    ['RD139_Narrow_UPS1_0_1fmol_inj1.mzML'],
                    file(params.modules_testdata_base_path + 'proteomics/diann/empirical_library.speclib', checkIfExists: true),
                    [file(params.modules_testdata_base_path + 'proteomics/diann/individual_analysis.quant', checkIfExists: true)],
                    file(params.modules_testdata_base_path + 'proteomics/diann/REF_EColi_K12_UPS1_combined_subset_100.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
