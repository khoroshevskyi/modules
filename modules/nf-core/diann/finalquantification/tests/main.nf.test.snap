{
    "proteomics - final quantification": {
        "content": [
            [
                [
                    {
                        "experiment_id": "test_exp"
                    },
                    "diann_report.stats.tsv:md5,a960024b8ef9e23c8010bfcd93e6df5a"
                ]
            ],
            [
                [
                    {
                        "experiment_id": "test_exp"
                    },
                    "diann_report.pr_matrix.tsv:md5,27e4024c6ea63ae0d82aabf11a55e790"
                ]
            ],
            [
                [
                    {
                        "experiment_id": "test_exp"
                    },
                    "diann_report.pg_matrix.tsv:md5,858750282171ae4d6fe3c361b1b3e42e"
                ]
            ],
            [
                "versions.yml:md5,4054b1f3f2e35dce5ba5cf698f2b9ce5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-24T14:50:29.686627"
    },
    "proteomics - final quantification - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "10": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "empirical_library.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "11": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "empirical_library.tsv.skyline.speclib:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "12": [
                    "versions.yml:md5,4054b1f3f2e35dce5ba5cf698f2b9ce5"
                ],
                "2": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.manifest.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.protein_description.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.stats.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.pr_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.pg_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.gg_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "8": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.unique_genes_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "9": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diannsummary.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "final_speclib": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "empirical_library.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "gg_matrix": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.gg_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diannsummary.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "main_report": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "pg_matrix": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.pg_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "pr_matrix": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.pr_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "protein_description": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.protein_description.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "report_manifest": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.manifest.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "report_parquet": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "report_stats": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.stats.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "skyline_speclib": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "empirical_library.tsv.skyline.speclib:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "unique_gene_matrix": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "diann_report.unique_genes_matrix.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4054b1f3f2e35dce5ba5cf698f2b9ce5"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-24T14:50:34.425042"
    }
}